src/components/drawing/index.js

```diff
@@ -455,84 +473,119 @@ drawing.makeSelectedPointStyleFns = function(trace) {
     var smsIsDefined = sms !== undefined;
     var usmsIsDefined = usms !== undefined;
 
-    if(smsIsDefined || usmsIsDefined) {
-        out.sizeFn = function(d) {
-            var mrc = d.mrc;
+    if(Registry.traceIs(trace, 'symbols') && (smsIsDefined || usmsIsDefined)) {
+        out.selectedSizeFn = function(d) {
+            var base = d.mrc || ms / 2;
+
             if(d.selected) {
-                return smsIsDefined ? sms / 2 : mrc;
+                return smsIsDefined ? sms / 2 : base;
             } else {
-                return usmsIsDefined ? usms / 2 : mrc;
+                return usmsIsDefined ? usms / 2 : base;
             }
         };
     }
 
     return out;
 };
 
+drawing.makeSelectedTextStyleFns = function(trace) {
+    var out = {};
+
+    var selectedAttrs = trace.selected || {};
+    var unselectedAttrs = trace.unselected || {};
+
+    var textFont = trace.textfont || {};
+    var selectedTextFont = selectedAttrs.textfont || {};
+    var unselectedTextFont = unselectedAttrs.textfont || {};
+
+    var tc = textFont.color;
+    var stc = selectedTextFont.color;
+    var utc = unselectedTextFont.color;
+
+    out.selectedTextColorFn = function(d) {
+        var base = d.tc || tc;
+
+        if(d.selected) {
+            return stc || base;
+        } else {
+            if(utc) return utc;
+            else return stc ? base : Color.addOpacity(base, DESELECTDIM);
+        }
+    };
+
+    return out;
+};
+
 drawing.selectedPointStyle = function(s, trace) {
     if(!s.size() || !trace.selectedpoints) return;
 
     var fns = drawing.makeSelectedPointStyleFns(trace);
     var marker = trace.marker || {};
+    var seq = [];
 
-    s.each(function(d) {
-        var pt = d3.select(this);
-        var mo2 = fns.opacityFn(d);
-        if(mo2 !== undefined) pt.style('opacity', mo2);
-    });
+    if(fns.selectedOpacityFn) {
+        seq.push(function(pt, d) {
+            pt.style('opacity', fns.selectedOpacityFn(d));
+        });
+    }
 
-    if(fns.colorFn) {
-        s.each(function(d) {
-            var pt = d3.select(this);
-            var mc2 = fns.colorFn(d);
-            if(mc2) Color.fill(pt, mc2);
+    if(fns.selectedColorFn) {
+        seq.push(function(pt, d) {
+            Color.fill(pt, fns.selectedColorFn(d));
         });
     }
 
-    if(Registry.traceIs(trace, 'symbols') && fns.sizeFn) {
-        s.each(function(d) {
-            var pt = d3.select(this);
+    if(fns.selectedSizeFn) {
+        seq.push(function(pt, d) {
             var mx = d.mx || marker.symbol || 0;
-            var mrc2 = fns.sizeFn(d);
+            var mrc2 = fns.selectedSizeFn(d);
 
             pt.attr('d', makePointPath(drawing.symbolNumber(mx), mrc2));
 
-            // save for selectedTextStyle
+            // save for Drawing.selectedTextStyle
             d.mrc2 = mrc2;
         });
     }
+
+    s.each(function(d) {
+        for(var i = 0; i < seq.length; i++) {
+            seq[i](d3.select(this), d);
+        }
```

Comment:
interesting approach with `seq` - it's not obvious to me that this is meaningfully different from 3 `s.each` calls, or 3 `if(functionExists)` blocks inside one `s.each`, did it make a big difference to do it this way?

Reply:
d3's each are a little slower than regular forEach and for-loops. Nothing worth losing sleep over, but we do gains a few ms using this sequence thing.

Reply:
... by gain, I mean lose a few ms and gain performance. :smile: